apps/wpcom-block-editor/src/wpcom/features/utils.js

```diff
@@ -223,19 +223,66 @@ const buildGlobalStylesEventProps = ( keyMap, value ) => {
 };
 
 /**
- * Builds and sends tracks events for global styles changes.
+ * Creates an array from all object properties, recursively.
+ *
+ * @param {Object} obj	Any object.
+ * @returns {Array}		Array of all object properties.
+ */
+function propertiesToArray( obj ) {
+	const isObject = ( val ) => val && typeof val === 'object' && ! Array.isArray( val );
+	const addDelimiter = ( a, b ) => ( a ? `${ a }.${ b }` : b );
+	const paths = ( object = {}, head = '' ) => {
+		return Object.entries( object ).reduce( ( product, [ key, value ] ) => {
+			const fullPath = addDelimiter( head, key );
+			return isObject( value )
+				? product.concat( paths( value, fullPath ) )
+				: product.concat( fullPath );
+		}, [] );
+	};
+	return paths( obj );
+}
+
+/**
+ * Records Tracks events for global styles updates.
  *
  * @param {Object} updated   The updated global styles content object.
  * @param {Object} original	 The original global styles content object.
  * @param {string} eventName Name of the tracks event to send.
+ * @returns {void}
  */
-export const buildGlobalStylesContentEvents = debounce( ( updated, original, eventName ) => {
-	// Debouncing is necessary to avoid spamming tracks events with updates when sliding inputs
-	// such as a color picker are in use.
-	return findUpdates( updated, original )?.forEach( ( { keyMap, value } ) => {
+const trackGlobalStylesEvents = ( updated, original, eventName ) => {
+	findUpdates( updated, original )?.forEach( ( { keyMap, value } ) => {
 		tracksRecordEvent( eventName, buildGlobalStylesEventProps( keyMap, value ) );
 	} );
-}, 100 );
+};
+
+/**
+ * Debounce wrapper for trackGlobalStylesEvent().
+ */
+const debounceTrackGlobalStylesEvents = debounce( trackGlobalStylesEvents, 300 );
+
+/**
+ * Builds and sends tracks events for global styles changes.
+ *
+ * @param {Object} updated   The updated global styles content object.
+ * @param {Object} original	 The original global styles content object.
+ * @param {string} eventName Name of the tracks event to send.
+ */
+export const buildGlobalStylesContentEvents = ( updated, original, eventName ) => {
+	const originalProperties = propertiesToArray( original );
+	const updatedProperties = propertiesToArray( updated );
+
+	// Debouncing is necessary to avoid spamming tracks events when
+	// using continuous controls like sliders or color pickers.
+	// We do it conditionally so we only debounce recurring updates
+	// to the same controls/styles. Some controls update multiple
+	// styles at once, in which case we do not want to debounce.
+	if ( isEqual( originalProperties, updatedProperties ) ) {
+		debounceTrackGlobalStylesEvents( updated, original, eventName );
```

Comment:
I think problem here is that once properties are set, this could always return `true` when changing them.

For example, (don't quote me on the exact structure here, just making something up for the purpose of explaining) a section of the global styles object might start off as:
```
typography: {
     fontSize: 5,
}
```
Other values will get populated into the `typography` object when they are set. When we set "appearance" for bold/italic, that could end up as:
```
typography: {
     fontSize: 5,
     fontWeight: 200,
     fontStyle: italic,
}
```
Now, imagine we set 'appearance' again that removes italic and changes the font weight. The final typography object will now look like:
```
typography: {
     fontSize: 5,
     fontWeight: 500,
     fontStyle: normal,
}
```
Now consider the above as `updated` and the previous example as `original`. The `isEqual` check on present properties should return `true` and it goes down the debouncing route, thus missing an update for one of the properties that changed.

Reply:
I hate to jump in and hack at this as I trust your ability to find a good solution. However my thoughts currently:

* This is a bit complicated to solve in a robust way and will take some time and thought.
* I think we have got to the point where we have learned everything we are going to codebase wise about this - so the onboarding benefit has been reached.
* There are probably more pressing bugs and other priorities to move into soon.
* I have had the time to let this problem/solution marinade in my head since writing the issue, and am pretty sure I have a robust solution in mind and have been itching to put it to code to see if it works.

[This is roughly my idea](https://github.com/Automattic/wp-calypso/pull/66049/files) about custom debouncing. In this approach the debouncing logic always runs, but we save `original` from the first call to compare to the `updated` of the last call so that we are sure to find all changes made to the GS object over those multiple/subsequent action calls. I tested briefly and it seems to work.

Curious about your thoughts here? If that approach seems to make sense you are welcome to implement something similar here, or clean that draft up and wrangle it over the line. If it doesn't maybe we move on to other issues for the time being.

Reply:
@Addison-Stavlo - No worries at all for jumping in here! Your example scenario above is a good one and it is correct - I just tested to confirm and my proposed solution still debounces when it shouldn't. 

And the solution you're proposing in the other PR seems correct logically to me (lots of nuanced logic there). I also just tested and it seems to work well in Chrome. 

I am seeing a small issue in Safari/Firefox. When sliding a color picker (try with Global Styles > Colors > Background Color), updates don't appear to be debounced. From quick checking, this seems to be because recurring calls to update global styles are happening at an interval just larger than the 300ms debounce time (around 400-500ms). 

In any case, I'd propose closing this PR, and moving the convo to your PR if that works for you.